### ISSUE TITLE
Simplify TestSource Class

### ISSUE THREAD

**Issue by jlink** [junit-team/junit5#175]
`TestSource` with its two concrete implementations is difficult to use since a cast to one of the implementations is required. I suggest to merge the interface and the two subclasses into a single interface with subclasses as needed along the following lines:
```
public interface TestSource extends Serializable {
   SourceType getType();
   Optional<Class<?>> getJavaClass();
   Optional<Method> getJavaMethod();
   Optional<Path> getFileOrDirectory();
   Optional<FilePosition> getFilePosition();
   enum SourceType { JavaClass, JavaMethod, Directory, File, FilePosition }
}
```

**Comment by bechte** [comment]
Two questions:

1. Is the enum required if we have all information in one big interface? It smells like a duplication of the class hierarchy below the interface.
2. If question 1 holds: we could simply remove the interface and make it a POJO, couldn't we? If so, I would like to have some Factory for it such than we can change or minds later without having to change all callers throughout the code base. 

How about the others?

**Comment by jlink** [comment]
@1: I'm not sure if the enum is strictly required, I _imagined_ it to simplify using it from the IDE's point of view but that's pure imagination. And it could simplify adding source types later like `URI` without clients having to check for new properties. I might be wrong, though.
@2: An interface is a looser contract but yes, a POJO might as good.

One point that I did not address in the issue description: The current implementation of `JavaSource` does not use `Method` but two props: `methodName` and `methodParameterTypes`. I guess that's for keeping serializability. Do we know that serializability is important here?

**Comment by marcphilipp** [comment]
Even though we only provide two implementations of `TestSource`, I'd like to keep the door open for third-party engines to provide their own implementation. Thus, I think having a static enum or a single implementation (POJO) would not be a good idea.

**Comment by jlink** [comment]
@marcphilipp We could add a value `Generic` to the enum and also provide a `Optional<Object> getGenericInfo()` property.

**Comment by marcphilipp** [comment]
In the current implementation, I don't think the `instanceof` makes it hard to use but rather that everything is optional. I don't think that will improve a whole lot by using the enum, will it?

**Comment by jlink** [comment]
Would having a specical subclass for all source types make it simpler? We
could then get rid of the optionals.

As for the enum: We currently have a few `isXXX` props and the enum seems
to be an improvement to that (and make the instanceof redundant).

2016-03-11 9:47 GMT+01:00 Marc Philipp <notifications@github.com>:

**Comment by marcphilipp** [comment]
### New proposal

<img width="801" alt="screen shot 2016-04-26 at 20 27 15" src="https://cloud.githubusercontent.com/assets/214207/14829759/6fe58cce-0bed-11e6-978f-91b9026fbfff.png">

@junit-team/junit-lambda What do you think?

**Comment by mmerdes** [comment]
The type hierarchy sure looks plausible.
(I would assume that for the purpose at hand it would _not_ make sense to model JavaSource as a subtype of FileSource)
@marcphilipp Do you oversee the implications for the implementation, especially for instanceOf()s?

**Comment by marcphilipp** [comment]
Feedback from team discussion: Use `Path` instead of `File`.

**Comment by marcphilipp** [comment]
### Updated proposal

<img width="1177" alt="screen shot 2016-04-27 at 17 03 36" src="https://cloud.githubusercontent.com/assets/214207/14856842/04ab46da-0c9a-11e6-96dd-229dc265ea1c.png">

**Comment by bechte** [comment]
I've started working on this issue.

**Comment by jlink** [comment]
You might also consider to add a JavaPackageSource and a UrlSource or UriSource (for stuff like Fitnesse tests that live in a web server).

**Comment by sbrannen** [comment]
Good points!

We actually discussed introducing a `JavaPackageSource`, but it didn't make it into the diagram.

A `UriSource` is also a very good idea. Since files and directories also have an associated URI, I would suggest that `UriSource` be an immediate superinterface of `FileSystemSource`.

**Comment by marcphilipp** [comment]
Like this?

<img width="965" alt="screen shot 2016-05-10 at 14 03 20" src="https://cloud.githubusercontent.com/assets/214207/15145790/fd726682-16b7-11e6-85f0-598556a455e5.png">

**Comment by mmerdes** [comment]
Looks good to me.
(I assume the return type of JavaPackageSource.getPackageName() should be a String rather than a Class?)

**Comment by marcphilipp** [comment]
Right. 😳

**Comment by sbrannen** [comment]
Yep, exactly like that (with the suggested change to `String` from Matthias). 😉

**Comment by bechte** [comment]
Merged with #255 into master.
